packages/dataviews/src/normalize-fields.ts

```diff
@@ -18,7 +18,9 @@ export function normalizeFields< Item >(
 
 		const getValue =
 			field.getValue ||
-			( ( { item }: { item: ItemRecord } ) => item[ field.id ] );
+			( ( { item }: { item: ItemRecord } ) =>
+				// @ts-ignore
```

Comment:
This is needed because I had to change the "ItemRecord" type from Record to Object. For me it should be possible to access object properties using a string key like here, so maybe there's another typescript friendly way to do it.

If I don't update the type of `ItemRecord` than the `orderField` definition becomes invalid requiring me to have an explicit `getValue` instead.

cc @jsnajdr in case you have thoughts.

Reply:
Interesting, for some reason `PostWithPageAttributesSupport` type is not assignable to `Record<string, unknown>`.

Reply:
This can be fixed by making `PostWithPageAttributesSupport` a `type` instead of `interface`. Because interfaces cannot be indexed by any string. Try this:
```ts
interface IPostWithPageAttributesSupport {
  menu_order: number;
}

type TPostWithPageAttributesSupport = {
  menu_order: number;
}

type IExt = IPostWithPageAttributesSupport extends Record<string,unknown> ? 'yes' : 'no';
type TExt = TPostWithPageAttributesSupport extends Record<string,unknown> ? 'yes' : 'no';
```
The `IExt` type is going to be `no` and `TExt` will be `yes`.

I experimented with making `ItemRecord` a bit smarter, to index on specific property instead of all strings:
```ts
type ItemRecord<Name extends string> = { [name in Name]: unknown };
type Field<Item, Name extends string = string> = { id: Name } & ( Item extends ItemRecord<Name> ? ... : ... );
const orderField: Field< PostWithPageAttributesSupport, 'menu_order' > = ...;
```
but this created an avalanche of other problems I was unable to fix.

FYI @sirreal as this is another interesting TS puzzle 🙂

Reply:
To be honest I wonder which is better. Keep that ts-ignore and support interfaces or force users of DataViews to use "type"

Reply:
It's very strange behavior indeed:

```ts
interface IPostWithPageAttributesSupport {
  menu_order: number;
  // [K: string]: unknown;
}

type TPostWithPageAttributesSupport = {
  menu_order: number;
}

// Depends on the string index in the interface.
type IExt = IPostWithPageAttributesSupport extends Record<string,unknown> ? 'yes' : 'no';

// "yes"
type TExt = TPostWithPageAttributesSupport extends Record<string,unknown> ? 'yes' : 'no';

declare var i: IPostWithPageAttributesSupport;
declare var t: TPostWithPageAttributesSupport;

// This is an error if the interface does not have the string index.
i['foo'];

// This is an error… but the type extends Record<string,unknown>???
t['foo'];
```

Reply:
> ```ts
> // This is an error… but the type extends Record<string,unknown>???
> t['foo'];
> ```

Interesting that this can be fixed with:
```ts
(t as Record<string, unknown>)['foo'];
```
but the `i['foo']` construct cannot be fixed because then TS complains that the two types don't sufficiently overlap for `as` to work.

Also you can assign `t` to a variable:
```ts
const rt: Record<string,unknown> = t;
```
but you cannot do that with `i`.

So, the fact that `A extends B` apparently means only that A is _assignable_ to B, but not that A can be directly used as B without any cast.

Reply:
All these type issues are caused by trying to be smart about the `Field.getValue` method. We try to figure out automatically whether it must be supplied by the caller, or whether it's optional and can be defaulted to:
```js
getValue = ( { item } => item[ field.id ] )
```
This works only for certain types of `item`, namely those that have a field called `field.id`. Or for those that can be at least safely used with the `[]` operator. Then the value will be `undefined`, but at least it won't crash, like e.g. `( null )[ 'foo' ]` would do.

If we stopped being smart about this, and made `getValue` always optional, then we could get rid of this complexity. Even at the cost of allowing a runtime crash when the `item` type is not indexable. But I'd say that in practice this is very unlikely.

Reply:
I'm implementing the suggestion from the previous comment in #64367.